fix(api): swallow hard-halt-related errors and hold lock in session.stop()

sfoster1 opened the pull request:
- Don't raise smoothieerrors for smoothie return strings like "alarm lock" or "after halt you should home" because they only happen when there's another previous error, and since they happen in an exception context they end up replacing the original SmoothieAlarm exception with a SmoothieError, which can be caught in many more circumstances, like in fast_home.
- Hold the motion lock for the call to hardware control stop(), to make sure that it doesn't happen until the call to session.run() has ended. This prevents a race condition where the halt() and stop() calls could happen in between the protocol run's smoothie calls, meaning that the
smoothie error response would be caught by the reset command instead of the protocol. The subsequent command would just look like a warning, so it wouldn't bubble up and the protocol wouldn't be cancelled.

Closes #4979


## Testing
- Cancel protocols. Try it on both API v1 and API v2; try it when waiting for a module command; try to get it right after the tip is dropped and the ejector shroud starts moving up.

sfoster1 committed:
fix(api): hold motion lock in session for calling stop()

This prevents a race condition where the halt() and stop() calls could
happen in between the protocol run's smoothie calls, meaning that the
smoothie error response would be caught by the reset command instead of
the protocol. The subsequent command would just look like a warning, so
it wouldn't bubble up and the protocol wouldn't be cancelled.

Closes #4979

nusrat813 commented:
Tested with module protocol on apiv1 and apiv2 - cancel is working properly right after drop tip

nusrat813 reviewed (approved):
Tested with module protocol on apiv1 and apiv2 - cancel is working properly right after drop tip

ahiuchingau reviewed (changes requested):
Tried cancelling during drop tip and the protocol (APIv2) was still not cancelling properly

nusrat813 commented:
Approval based on instructions provided - "try to get it right after the tip is dropped"

sfoster1 committed:
further fix: don't swallow alarms

This could have been affecting other pause cases. There's many times in
the system where we wrap motion commands in try...finally to clean up.
Those cleanups usually involve sending other commands.

If you just did a smoothie hard halt, then the smoothie will respond to
those commands with "alarm lock" until you send an m999.

These will raise SmoothieErrors. Since this is in an exception context
it will put a new exception on the exception stack. That new exception
will be a smoothie error, and _that_ may be handled by code like fast
home and the cancel may never happen.

Now, we actually check to see if the error is an alarm lock or the
warning about homing and don't raise an error for them - there's already
been an error, the original kill button pressed one.

nusrat813 commented:
Changes tested while drop tip in montion for apiv1 and apiv2 - cancel works

sfoster1 committed:
fixup: line length

ahiuchingau reviewed (approved):
The fix worked on v1 & v2 protocols on 🌆 & code looks good to me
